leapp/utils/lock.py

```diff
@@ -0,0 +1,98 @@
+import os
+import fcntl
+import logging
+
+from leapp.config import get_config
+from leapp.exceptions import ProcessLockError
+
+
+def leapp_lock(lockfile=None):
+    return ProcessLock(lockfile=lockfile)
+
+
+def _acquire_lock(fd):
+    try:
+        fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        return True
+    except OSError:
+        return False
+
+
+def _read_pid(fd):
+    return os.read(fd, 20)
+
+
+def _write_pid(fd, pid):
+    os.write(fd, str(pid).encode('utf-8'))
+
+
+def _validate_pid(old_pid):
+    try:
+        old_pid = int(old_pid)
+    except ValueError:
+        return False
+
+    return old_pid
+
+
+def _is_old_process_running(old_pid):
+    return os.access('/proc/{}/stat'.format(old_pid), os.F_OK)
+
+
+def _clear_lock(fd):
+    os.lseek(fd, 0, os.SEEK_SET)
+    os.ftruncate(fd, 0)
+
+
+class ProcessLock(object):
+
+    def __init__(self, lockfile=None):
+        self.log = logging.getLogger('leapp.utils.lock')
+        self.lockfile = lockfile if lockfile else get_config().get('lock', 'path')
+
+    def _try_lock(self, pid):
+        fd = os.open(self.lockfile, os.O_CREAT | os.O_RDWR, 0o644)
+
+        try:
+            if not _acquire_lock(fd):
+                msg = ('Leapp is currently locked by another process and cannot be started.\n'
+                       'Please ensure all previous instances of the application are closed and try again.')
+                raise ProcessLockError(msg)
+
+            old_pid = _read_pid(fd)
+            if len(old_pid) == 0:
+                # No pid in lockfile
+                _write_pid(fd, pid)
+                return pid
+
+            old_pid = _validate_pid(old_pid)
```

Comment:
I was checking some corner cases: currently if you modify the PID stored in leapp file a parallel leapp run execution would be allowed, and in the end both runs will be terminated with the following trace
```
Traceback (most recent call last):
  File "/usr/bin/leapp", line 11, in <module>
    load_entry_point('leapp==0.17.0', 'console_scripts', 'leapp')()
  File "/usr/lib/python3.6/site-packages/leapp/cli/__init__.py", line 46, in main
    cli.command.execute('leapp version {}'.format(VERSION))
  File "/usr/lib/python3.6/site-packages/leapp/utils/lock.py", line 98, in __exit__
    os.unlink(self.lockfile)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/leapp/leapp_lock.pid'
```

Maybe we could change that? To me it feels like if there is already a lock file present regardless of contents - no other leapp run should be allowed at least. WDYT?

Reply:
I see your point. The reason the PID is stored within the lock file is that leapp could have been terminated with a SIGKILL previously, which would not remove the lock file, but the idea here was that leapp is not in fact running and should start normally.

What you are suggesting is, I am guessing, that leapp would not run in this case and we would prompt the user to delete the lock file manually.

The main disadvantage I see that in case someone creates the lock file after running ``leapp upgrade`` but before reboot (suppose we take this into account as we are already considering someone externally modifying the lock file during execution; the current mechanism should not corrupt the file), leapp would fail in initram beause it detects an existing lock file. The current approach would at least check whether a process with the PID in it is running and proceed.

Overall I think there is not much we can guarantee if someone externally modifies the lock, but the way it is there is at least less of a chance we fail in initram because of it, so I would propose to keep it as is.

Reply:
If someone touches leapp internals (e.g. modifies the lock file, or creates a lock file after execution) then it is their problem what happens (e.g. leapp fails in initramfs). Although these concerns can be hypothetically materialized, I do not think that we should spend time discussing them.

As for leaving out a PID file after a SIGKILL -- same applies. The user has opted in to killing leapp in a way that it cannot clean after itself. So using SIGKILL semantically involves acknowledging the fact that there will be mess on the system. As long as the lock file can be removed manually afterwards, everything is fine.

Reply:
Although I agree with modification of leapp internals pushing responsibility onto the user, I think dealing with a stale lockfile in the middle of an upgrade, is important.  Bugs, things outside of the user's conscious control on the system, etc can lead to stale lockfiles.  While it's somewhat standard for users to have to clean up those lockfiles if necessary when invoking a program manually from the command line, forcing the user to clean up a stale lockfile themselves while leapp is supposed to be handling the next steps of the upgrade automatically feels like it will cause the user quite a bit of anxiety that it would be best for us to avoid.